egs/librispeech/WSASR/README.md

```diff
@@ -0,0 +1,204 @@
+
+
+We modify $G(\mathbf{y})$ by adding self-loop arcs into each state and bypass arcs into each arc. 
+  <p align="center">
+    <img src="figures/otc_g.png" alt="Image Alt Text" width="50%" />
+
+  </p>
+
+After composing the modified WFST $G_{\text{otc}}(\mathbf{y})$ with $L$ and $T$, the OTC training graph is shown in this figure:
+<figure style="text-align: center">
+  <img src="figures/otc_training_graph.drawio.png" alt="Image Alt Text" />
+  <figcaption>OTC training graph. The self-loop arcs and bypass arcs are highlighted in green and blue, respectively.</figcaption>
+</figure>
+
+The $\star$ is represented as the average probability of all non-blank tokens.
+  <p align="center">
+    <img src="figures/otc_emission.drawio.png" width="50%" />
+  </p>
+
+The weight of $\star$ is the log average probability of "a" and "b": $\log \frac{e^{-1.2} + e^{-2.3}}{2} = -1.6$ and $\log \frac{e^{-1.9} + e^{-0.5}}{2} = -1.0$ for 2 frames.
+
+## Description of the recipe
+### Preparation
+```
+# feature_type can be ssl or fbank
+feature_type=ssl
+feature_dir="data/${feature_type}"
+manifest_dir="${feature_dir}"
+lang_dir="data/lang"
+lm_dir="data/lm"
+exp_dir="conformer_ctc2/exp"
+otc_token="<star>"
+
+./prepare.sh \
+  --feature-type "${feature_type}" \
+  --feature-dir "${feature_dir}" \
+  --lang-dir "${lang_dir}" \
+  --lm-dir "${lm_dir}" \
+  --otc-token "${otc_token}" 
+```
+This script adds the 'otc_token' ('\<star\>') and its corresponding sentence-piece ('▁\<star\>') to 'words.txt' and 'tokens.txt,' respectively. Additionally, it computes SSL features using the 'wav2vec2-base' model. (You can use GPU to accelerate feature extraction).
+
+### Making synthetic errors to the transcript [optional]
+```
+sub_er=0.17
+ins_er=0.17
+del_er=0.17
+synthetic_train_manifest="librispeech_cuts_train-clean-100_${sub_er}_${ins_er}_${del_er}.jsonl.gz"
+
+./local/make_error_cutset.py \
+  --input-cutset "${manifest_dir}/librispeech_cuts_train-clean-100.jsonl.gz" \
+  --words-file "${lang_dir}/words.txt" \
+  --sub-error-rate "${sub_er}" \
+  --ins-error-rate "${ins_er}" \
+  --del-error-rate "${del_er}" \
+  --output-cutset "${manifest_dir}/${synthetic_train_manifest}"
+```
+This script generates synthetic substitution, insertion, and deletion errors in the transcript with ratios 'sub_er', 'ins_er', and 'del_er', respectively. The original transcript is saved as 'verbatim transcript' in the cutset, along with information on how the transcript is corrupted:
+  - '[hello]' indicates the original word 'hello' is substituted by another word
+  - '[]' indicates an extra word is inserted into the transcript
+  - '-hello-' indicates the word 'hello' is deleted from the transcript
+So if the original transcript is "have a nice day" and the synthetic one is "a very good day", the 'verbatim transcript' would be:
+```
+original:  have  a      nice  day
+synthetic:       a very good  day
+verbatim: -have- a  [] [nice] day
+```
+
+### Training
```

Comment:
Please put a note that only train-clean-100 is used for training and not the full librispeech dataset.

Reply:
Sure.